README.md
<div align="center" style="text-align:center">
    
<img height="125" src="https://b.catgirlsare.sexy/WcjHI5SQ.png" />

[![Visual Studio Marketplace Version](https://img.shields.io/visual-studio-marketplace/v/Kurozero.vscode-commit-reminder)](https://marketplace.visualstudio.com/items?itemName=Kurozero.vscode-commit-reminder)
[![Visual Studio Marketplace Rating (Stars)](https://img.shields.io/visual-studio-marketplace/stars/Kurozero.vscode-commit-reminder)](https://marketplace.visualstudio.com/items?itemName=Kurozero.vscode-commit-reminder)
[![Visual Studio Marketplace Downloads](https://img.shields.io/visual-studio-marketplace/d/Kurozero.vscode-commit-reminder?label=downloads)](https://marketplace.visualstudio.com/items?itemName=Kurozero.vscode-commit-reminder)

# VSCode Commit Reminder
Periodically notify to commit new changes

# Support
[![discord](https://discordapp.com/api/v6/guilds/240059867744698368/widget.png?style=banner2)](https://discord.gg/p895czC)

</div>

CHANGELOG.md
# Change Log
All notable changes to the "vscode-commit-reminder" extension will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.0] Oct 26, 2020
### Added
    - New config option intervalUnit, defaults to minutes
### Changed
    - Renamed output channel from VSCode Commit Reminder > Commit Reminder
### Misc
    - Updated dependencies

## [0.2.1] Aug 20, 2020
### Misc
    - Updated dependencies

## [0.2.0] Jul 4, 2020
### Added
    - New config option minimumFileChanges, defaults to 2
    - New config option minimumChanges, defaults to 10 (insertions + deletions)
### Changed
    - Moved error logs to output channel
### Misc
    - Updated dependencies

## [0.1.1] Jun 19, 2020
### Added
    - Added this changelog
### Changed
    - Bump dependency to fix initial interval message

## [0.1.0] Jun 19, 2020
### Fixed
    - Fixed interval minutes from last update

## [0.0.3] Jun 19, 2020
### Added
    - Check for git repo when config is changed
### Changed
    - rootPath to workspaceFolders
### ⚠️ Warning
    - Do NOT use this version since I messed the configurable minutes up and will spam messages every 5 seconds

## [0.0.2] Jun 17, 2020
### Changed
    - Updated dependencies
    - Minimum vscode version from 1.40.0 to 1.46.0
### Fixed
    - Check if workspace is a git repo
    - Moved error message from popup to console

## [0.0.1] Dec 13, 2019
### Added
    - Initial Release


.eslintrc.js
module.exports = {
    "plugins": [
        "@typescript-eslint"
    ],
    "env": {
        "es6": true,
        "node": true
    },
    "extends": [
        "plugin:@typescript-eslint/recommended"
    ],
    "parserOptions": {
        "ecmaVersion": 10,
        "parser": "@typescript-eslint/parser"
    },
    "rules": {
        "@typescript-eslint/explicit-function-return-type": ["error", {
            "allowExpressions": true
        }],
        "@typescript-eslint/interface-name-prefix": "off",
        "@typescript-eslint/no-explicit-any": "off",
        "@typescript-eslint/no-non-null-assertion": "off",
        "accessor-pairs": "error",
        "array-bracket-newline": "error",
        "array-bracket-spacing": [
            "error",
            "never"
        ],
        "array-callback-return": "error",
        "array-element-newline": "off",
        "arrow-body-style": "error",
        "arrow-parens": [
            "error",
            "always"
        ],
        "arrow-spacing": [
            "error",
            {
                "after": true,
                "before": true
            }
        ],
        "block-scoped-var": "error",
        "block-spacing": [
            "error",
            "always"
        ],
        "brace-style": [
            "error",
            "1tbs",
            {
                "allowSingleLine": true
            }
        ],
        "callback-return": "off",
        "capitalized-comments": "off",
        "class-methods-use-this": "off",
        "comma-dangle": "error",
        "comma-spacing": [
            "error",
            {
                "after": true,
                "before": false
            }
        ],
        "comma-style": [
            "error",
            "last"
        ],
        "complexity": [
            "error",
            30
        ],
        "computed-property-spacing": [
            "error",
            "never"
        ],
        "consistent-return": "off",
        "consistent-this": "error",
        "curly": "off",
        "default-case": "error",
        "dot-location": "error",
        "dot-notation": [
            "error",
            {
                "allowKeywords": true
            }
        ],
        "eol-last": "error",
        "eqeqeq": "error",
        "func-call-spacing": "error",
        "func-name-matching": "error",
        "func-names": "off",
        "func-style": [
            "error",
            "declaration"
        ],
        "function-paren-newline": "error",
        "generator-star-spacing": "error",
        "global-require": "off",
        "guard-for-in": "off",
        "handle-callback-err": "error",
        "id-blacklist": "error",
        "id-length": "off",
        "id-match": "error",
        "implicit-arrow-linebreak": [
            "error",
            "beside"
        ],
        "indent": "off",
        "indent-legacy": "off",
        "init-declarations": "error",
        "jsx-quotes": "error",
        "key-spacing": "error",
        "keyword-spacing": [
            "error",
            {
                "after": true,
                "before": true
            }
        ],
        "line-comment-position": "off",
        "linebreak-style": [
            "error",
            "unix"
        ],
        "lines-around-comment": "off",
        "lines-around-directive": "error",
        "lines-between-class-members": ["error", "always", {
            "exceptAfterSingleLine": true
        }],
        "max-classes-per-file": [
            "error",
            5
        ],
        "max-depth": [
            "error",
            10
        ],
        "max-len": "off",
        "max-lines": "error",
        "max-lines-per-function": [
            "error",
            150
        ],
        "max-nested-callbacks": "error",
        "max-params": "off",
        "max-statements": [
            "error",
            50
        ],
        "max-statements-per-line": "off",
        "multiline-comment-style": [
            "error",
            "separate-lines"
        ],
        "multiline-ternary": "off",
        "new-cap": "off",
        "new-parens": "error",
        "newline-after-var": "off",
        "newline-before-return": "off",
        "newline-per-chained-call": "off",
        "no-alert": "error",
        "no-array-constructor": "error",
        "no-async-promise-executor": "error",
        "no-await-in-loop": "off",
        "no-bitwise": "error",
        "no-buffer-constructor": "error",
        "no-caller": "error",
        "no-catch-shadow": "error",
        "no-confusing-arrow": "off",
        "no-console": "off",
        "no-continue": "off",
        "no-div-regex": "error",
        "no-duplicate-imports": "error",
        "no-else-return": "off",
        "no-empty-function": "off",
        "no-eq-null": "error",
        "no-eval": "error",
        "no-extend-native": "error",
        "no-extra-bind": "error",
        "no-extra-label": "error",
        "no-extra-parens": "off",
        "no-floating-decimal": "error",
        "no-implicit-coercion": "error",
        "no-implicit-globals": "error",
        "no-implied-eval": "error",
        "no-inline-comments": "off",
        "no-invalid-this": "error",
        "no-iterator": "error",
        "no-label-var": "error",
        "no-labels": "error",
        "no-lone-blocks": "error",
        "no-lonely-if": "off",
        "no-loop-func": "error",
        "no-magic-numbers": "off",
        "no-misleading-character-class": "error",
        "no-mixed-operators": "error",
        "no-mixed-requires": "error",
        "no-multi-assign": "error",
        "no-multi-spaces": "error",
        "no-multi-str": "error",
        "no-multiple-empty-lines": "error",
        "no-native-reassign": "error",
        "no-negated-condition": "off",
        "no-negated-in-lhs": "error",
        "no-nested-ternary": "off",
        "no-new": "error",
        "no-new-func": "error",
        "no-new-object": "error",
        "no-new-require": "error",
        "no-new-wrappers": "error",
        "no-octal-escape": "error",
        "no-param-reassign": "error",
        "no-path-concat": "error",
        "no-plusplus": "off",
        "no-process-env": "off",
        "no-process-exit": "off",
        "no-proto": "error",
        "no-prototype-builtins": "off",
        "no-restricted-globals": "error",
        "no-restricted-imports": "error",
        "no-restricted-modules": "error",
        "no-restricted-properties": "error",
        "no-restricted-syntax": "error",
        "no-return-assign": "off",
        "no-return-await": "off",
        "no-script-url": "error",
        "no-self-compare": "error",
        "no-sequences": "error",
        "no-shadow": "off",
        "no-shadow-restricted-names": "error",
        "no-spaced-func": "error",
        "no-sync": [
            "error",
            {
                "allowAtRootLevel": true,
            }
        ],
        "no-tabs": "error",
        "no-template-curly-in-string": "error",
        "no-ternary": "off",
        "no-throw-literal": "error",
        "no-trailing-spaces": "error",
        "no-undef-init": "error",
        "no-undefined": "off",
        "no-underscore-dangle": "off",
        "no-unmodified-loop-condition": "error",
        "no-unneeded-ternary": "error",
        "no-unused-expressions": "error",
        "no-use-before-define": "error",
        "no-useless-call": "error",
        "no-useless-catch": "off",
        "no-useless-computed-key": "error",
        "no-useless-concat": "error",
        "no-useless-constructor": "error",
        "no-useless-rename": "error",
        "no-useless-return": "error",
        "no-var": "error",
        "no-void": "off",
        "no-warning-comments": "off",
        "no-whitespace-before-property": "error",
        "no-with": "error",
        "nonblock-statement-body-position": [
            "error",
            "any"
        ],
        "object-curly-newline": "error",
        "object-curly-spacing": [
            "error",
            "always"
        ],
        "object-shorthand": "off",
        "one-var": "off",
        "one-var-declaration-per-line": "error",
        "operator-assignment": "error",
        "operator-linebreak": "error",
        "padded-blocks": "off",
        "padding-line-between-statements": "error",
        "prefer-arrow-callback": "error",
        "prefer-const": "off",
        "prefer-destructuring": "off",
        "prefer-numeric-literals": "error",
        "prefer-object-spread": "error",
        "prefer-promise-reject-errors": "error",
        "prefer-reflect": "error",
        "prefer-rest-params": "error",
        "prefer-spread": "error",
        "prefer-template": "error",
        "quote-props": "off",
        "quotes": [
            "error",
            "double"
        ],
        "radix": "error",
        "require-atomic-updates": "error",
        "require-await": "off",
        "require-jsdoc": "off",
        "require-unicode-regexp": "error",
        "rest-spread-spacing": "error",
        "semi": "error",
        "semi-spacing": "error",
        "semi-style": [
            "error",
            "last"
        ],
        "sort-imports": "off",
        "sort-keys": "off",
        "sort-vars": "error",
        "space-before-blocks": "error",
        "space-before-function-paren": "off",
        "space-in-parens": [
            "error",
            "never"
        ],
        "space-infix-ops": "error",
        "space-unary-ops": [
            "error",
            {
                "nonwords": false,
                "words": true
            }
        ],
        "spaced-comment": [
            "error",
            "always"
        ],
        "strict": "error",
        "switch-colon-spacing": "error",
        "symbol-description": "error",
        "template-curly-spacing": [
            "error",
            "never"
        ],
        "template-tag-spacing": "error",
        "unicode-bom": [
            "error",
            "never"
        ],
        "valid-jsdoc": "off",
        "vars-on-top": "error",
        "wrap-iife": "error",
        "wrap-regex": "error",
        "yield-star-spacing": "error",
        "yoda": [
            "error",
            "never"
        ]
    }
};

out/extension.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const vscode = __importStar(require("vscode"));
const promise_1 = __importDefault(require("simple-git/promise"));
const yukikaze_1 = __importDefault(require("yukikaze"));
const git = promise_1.default((_a = vscode.workspace.rootPath, (_a !== null && _a !== void 0 ? _a : process.cwd())));
const interval = new yukikaze_1.default();
// this method is called when your extension is activated
function activate(context) {
    var _a;
    const config = vscode.workspace.getConfiguration();
    const minutes = (_a = config.get("commitReminder.interval"), (_a !== null && _a !== void 0 ? _a : 30));
    startInterval(minutes);
    context.subscriptions.push(vscode.workspace.onDidChangeConfiguration((e) => {
        var _a;
        if (e.affectsConfiguration("commitReminder.interval")) {
            interval.stop();
            const config = vscode.workspace.getConfiguration();
            const minutes = (_a = config.get("commitReminder.interval"), (_a !== null && _a !== void 0 ? _a : 30));
            startInterval(minutes);
        }
    }));
}
exports.activate = activate;
// this method is called when your extension is deactivated
function deactivate() {
    interval.stop();
}
exports.deactivate = deactivate;
function startInterval(minutes) {
    interval.run(() => __awaiter(this, void 0, void 0, function* () {
        try {
            const result = yield git.diffSummary();
            if (result.changed > 0) {
                vscode.window.showInformationMessage(`Don't forget to commit your new changes | ${result.changed} ${result.changed === 1 ? "file" : "files"} to commit`);
            }
        }
        catch (err) {
            vscode.window.showErrorMessage(err ? err.toString() : "Error");
        }
    }), 1000 * minutes); // 1000 * 60 * minutes
}
//# sourceMappingURL=extension.js.map

src/extension.ts
import vscode from "vscode";
import SimpleGit from "simple-git";
import Yukikaze from "yukikaze";
import { Config } from "./types";

const workspace = vscode.workspace.workspaceFolders && vscode.workspace.workspaceFolders.length > 0 ? vscode.workspace.workspaceFolders[0].uri.fsPath : process.cwd();

const git = SimpleGit(workspace);
const interval = new Yukikaze();

const logger = vscode.window.createOutputChannel("Commit Reminder");

/**
 * Returns the extensions config values
 */
function getConfig(): Config {
    const config = vscode.workspace.getConfiguration();
    const intervalUnit = config.get<string>("commitReminder.intervalUnit") || "minutes";
    const interval = config.get<number>("commitReminder.interval") ?? 30;
    const minFileChanges = config.get<number>("commitReminder.minimumFileChanges") ?? 2;
    const minChanges = config.get<number>("commitReminder.minimumChanges") ?? 10;
    return { intervalUnit, interval, minFileChanges, minChanges };
}

/**
 * Check for changes every x [seconds, minutes, hours]
 */
function startInterval(): void {
    const config = getConfig();

    let multiplier = 1000 * 60;
    switch (config.intervalUnit) {
        case "seconds":
            multiplier = 1000;
            break;
        case "hours":
            multiplier = 1000 * 60 * 60;
            break;
    }

    // prettier-ignore
    interval.run(async () => {
        try {
            const result = await git.diffSummary();
            const totalChanges = result.insertions + result.deletions;
            if (result.changed >= config.minFileChanges || totalChanges >= config.minChanges) {
                vscode.window.showInformationMessage(`Don't forget to commit your new changes | ${result.changed} ${result.changed === 1 ? "file" : "files"} to commit`);
            }
        } catch (err) {
            vscode.window.showErrorMessage(err ? err.toString() : "Error");
        }
    }, multiplier * config.interval, false);
}

/**
 * Check if workspace is a git repo
 *
 * @returns Whether the workspace is a repo
 */
async function checkIsRepo(): Promise<boolean> {
    try {
        const isRepo = await git.checkIsRepo();
        if (isRepo) {
            return true;
        } else {
            logger.appendLine("vscode-commit-reminder - no git repository found");
            return false;
        }
    } catch (error) {
        logger.appendLine(`vscode-commit-reminder - no git repository found: ${error.name} ${error.message}`);
        return false;
    }
}

/**
 * If interval minutes is changed stop and restart interval
 */
const onConfigurationChanged = vscode.workspace.onDidChangeConfiguration(async (event) => {
    const unitChanged = event.affectsConfiguration("commitReminder.intervalUnit");
    const intervalChanged = event.affectsConfiguration("commitReminder.interval");
    const minFilesChanged = event.affectsConfiguration("commitReminder.minimumFileChanges");
    const minChangesChanged = event.affectsConfiguration("commitReminder.minimumChanges");

    // Restart interval if any ofter config options are affected by the changes
    if (unitChanged || intervalChanged || minFilesChanged || minChangesChanged) {
        interval.stop();

        // Check if current workspace is a git repo
        const isRepo = await checkIsRepo();
        if (isRepo) {
            startInterval();
        }
    }
});

/**
 * Runs on extension activation
 *
 * @param context context of the extension
 */
export async function activate(context: vscode.ExtensionContext): Promise<void> {
    context.subscriptions.push(onConfigurationChanged); // Add config change listener

    // Check if current workspace is a git repo
    const isRepo = await checkIsRepo();
    if (isRepo) {
        startInterval();
    }
}

/**
 * Runs on extension deactivate
 */
export function deactivate(): void {
    interval.stop();
}
